pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -141,23 +141,13 @@ func NewController(client kubernetes.Interface, options ControllerOptions) *Cont
 			return client.CoreV1().Services(options.WatchedNamespace).Watch(opts)
 		})
 
-	if out.EDSUpdater != nil {
-		out.endpoints = out.createEDSInformer(&v1.Endpoints{}, "Endpoints", options.ResyncPeriod,
-			func(opts meta_v1.ListOptions) (runtime.Object, error) {
-				return client.CoreV1().Endpoints(options.WatchedNamespace).List(opts)
-			},
-			func(opts meta_v1.ListOptions) (watch.Interface, error) {
-				return client.CoreV1().Endpoints(options.WatchedNamespace).Watch(opts)
-			})
-	} else {
-		out.endpoints = out.createInformer(&v1.Endpoints{}, "Endpoints", options.ResyncPeriod,
-			func(opts meta_v1.ListOptions) (runtime.Object, error) {
-				return client.CoreV1().Endpoints(options.WatchedNamespace).List(opts)
-			},
-			func(opts meta_v1.ListOptions) (watch.Interface, error) {
-				return client.CoreV1().Endpoints(options.WatchedNamespace).Watch(opts)
-			})
-	}
+	out.endpoints = out.createEDSInformer(&v1.Endpoints{}, "Endpoints", options.ResyncPeriod,
```

Comment:
I don't think we can remove the other option - it is triggered by 'disable incremental EDS' which proved to be very useful... I think we only need to fix the 'incremental EDS' side - the other seems to recover ( it pushes often enough )

Reply:
I think it will still work. The only difference between createEDSInformer is that it will cast the config to a `v1.Endpoints` and compare the subsets to determine if it needs to be updated, rather than the full object. This fixes the issue as kube-system has a lot of updates to resourceVersion which isn't relevant to us.

The problem is, EDSUpdater is always `nil` here; it is later set here: https://github.com/istio/istio/blob/c76b662ff31b491bde60b1af2f1cdcae7ab3ff51/pilot/pkg/bootstrap/server.go#L793

So even if we using incremental EDS, we wouldn't be making an createEDSInformer. If we are not using incremental EDS, I don't see why we cannot still use createEDSInformer -- it will handle it differently here https://github.com/istio/istio/blob/be7ca45f932e7462833e573c050ce5654b7a45ac/pilot/pkg/serviceregistry/kube/controller.go#L776-L787

Reply:
Yes, I remember doing that optimization for the incremental path - but not on the old path, to minimize risks.

Since this is 1.0 and we can't do a full test cycle - we need to be extra careful. Technically the new code path (incremental) has been tested extensively in 1.1, so maybe it's no longer risky to use the new code. 

Let me think about it more and look at the code - my preference is to minimize risk, if what you really need is to apply the same optimization on the diff - maybe we can just copy that. 

I doubt EDSUpdater can be always nil here - it would mean the createEDSInformer is never called, and I don't see how incremental EDS would work otherwise.

Reply:
I don't think that incremental EDS would not work if createEDSInformer is never called. As far as I can tell, the only difference in createEDSInformer is the optimization, so it should still work fine, just not use the optimization.